airbyte-config/persistence/src/main/java/io/airbyte/config/persistence/ConfigRepository.java

```diff
@@ -615,35 +629,86 @@ public void updateConnectionState(final UUID connectionId, final State state) th
     return actorCatalogs;
   }
 
-  public void writeCatalog(final AirbyteCatalog catalog,
-                           final UUID sourceId,
-                           final String configurationHash,
-                           final String connectorVersion)
-      throws JsonValidationException, IOException {
+  public Map<UUID, AirbyteCatalog> findCatalogByHash(final String catalogHash) throws IOException {
+    final Result<Record2<UUID, JSONB>> records = database.transaction(ctx -> ctx.select(ACTOR_CATALOG.ID, ACTOR_CATALOG.CATALOG)
+        .from(ACTOR_CATALOG)
+        .where(ACTOR_CATALOG.CATALOG_HASH.eq(catalogHash))).fetch();
+
+    final Map<UUID, AirbyteCatalog> result = new HashMap<>();
+    for (final Record record : records) {
+      final AirbyteCatalog catalog = Jsons.deserialize(
+          record.get(ACTOR_CATALOG.CATALOG).toString(), AirbyteCatalog.class);
+      result.put(record.get(ACTOR_CATALOG.ID), catalog);
+    }
+    return result;
+  }
+
+  public Optional<AirbyteCatalog> getActorCatalog(final UUID actorId,
+                                                  final String actorVersion,
+                                                  final String configHash)
+      throws IOException {
+    final Result<Record1<JSONB>> records = database.transaction(ctx -> ctx.select(ACTOR_CATALOG.CATALOG)
+        .from(ACTOR_CATALOG).join(ACTOR_CATALOG_FETCH_EVENT)
+        .on(ACTOR_CATALOG.ID.eq(ACTOR_CATALOG_FETCH_EVENT.ACTOR_CATALOG_ID))
+        .where(ACTOR_CATALOG_FETCH_EVENT.ACTOR_ID.eq(actorId))
+        .and(ACTOR_CATALOG_FETCH_EVENT.ACTOR_VERSION.eq(actorVersion))
+        .and(ACTOR_CATALOG_FETCH_EVENT.CONFIG_HASH.eq(configHash))).fetch();
+
+    if (records.size() > 1) {
+      throw new IllegalStateException(
+          String.format("Multiple catalog stored for ActorId=%s,ActorVersion=%s,ConfigHash=%s", actorId.toString(), actorVersion, configHash));
+    } else if (records.size() == 1) {
+      final JSONB record = records.get(0).get(ACTOR_CATALOG.CATALOG);
+      return Optional.of(Jsons.deserialize(record.toString(), AirbyteCatalog.class));
+    }
+    return Optional.empty();
+
+  }
+
+  public UUID writeActorCatalog(final AirbyteCatalog airbyteCatalog) throws IOException {
+    final OffsetDateTime timestamp = OffsetDateTime.now();
+    final HashFunction hashFunction = Hashing.murmur3_32_fixed();
+    final UUID catalogId = UUID.randomUUID();
+    final String catalogHash = hashFunction.hashBytes(Jsons.serialize(airbyteCatalog).getBytes(
+        Charsets.UTF_8)).toString();
+    database.transaction(ctx -> ctx.insertInto(ACTOR_CATALOG)
+        .set(ACTOR_CATALOG.ID, catalogId)
+        .set(ACTOR_CATALOG.CATALOG, JSONB.valueOf(Jsons.serialize(airbyteCatalog)))
+        .set(ACTOR_CATALOG.CATALOG_HASH, catalogHash)
+        .set(ACTOR_CATALOG_FETCH_EVENT.CREATED_AT, timestamp)
+        .set(ACTOR_CATALOG_FETCH_EVENT.MODIFIED_AT, timestamp)).execute();
```

Comment:
should this be `ACTOR_CATALOG.CREATED_AT`?

Reply:
why did we need a modified at for this table?

Reply:
i was imagining if the same catalog was written twice, that we would not end up with a duplicate in this table. we would just not write the new catalog and point to the existing one. in other words, the `catalog_hash` would be a unique index.

how are you thinking about it?

Reply:
I agree that we should not overwrite catalog but create new entries as needed. I kept the modified_at, created_at columns as they are present on all table and the persistence class and the ConfigWithMetadata really expected those information. I do not expect catalog to be modified expect in rare case of an admin operation